pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ruler
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/rules"
+	"golang.org/x/sync/semaphore"
+)
+
+type MultiTenantRuleConcurrencyController interface {
+	rules.RuleConcurrencyController
+}
+
+const IntervalToEvaluationThreshold = 50.00
+
+// DynamicSemaphore is a semaphore that can dynamically change its max concurrency.
+// It is necessary as the max concurrency is defined by the user limits which can be changed at runtime.
+type DynamicSemaphore struct {
+	logger         log.Logger
+	maxConcurrency func() int64
+
+	mtx      sync.Mutex
+	acquired int64
+}
+
+// NewDynamicSemaphore creates a new DynamicSemaphore
+func NewDynamicSemaphore(logger log.Logger, maxConcurrency func() int64) *DynamicSemaphore {
+	return &DynamicSemaphore{
+		logger:         logger,
+		maxConcurrency: maxConcurrency,
+		acquired:       0,
+	}
+}
+
+// TryAcquire tries to acquire a token from the semaphore.
+func (ds *DynamicSemaphore) TryAcquire() bool {
+	ds.mtx.Lock()
+	defer ds.mtx.Unlock()
+	maxTokens := ds.maxConcurrency()
+	if ds.acquired < maxTokens {
+		ds.acquired++
+		return true
+	}
+
+	return false
+}
+
+// Release releases a token back to the semaphore.
+func (ds *DynamicSemaphore) Release() {
+	ds.mtx.Lock()
+	defer ds.mtx.Unlock()
+	if ds.acquired > 0 {
+		ds.acquired--
+	} else {
+		panic("dynamic semaphore: release called without corresponding acquire")
+	}
+}
+
+type MultiTenantConcurrencyControllerMetrics struct {
+	CurrentConcurrency prometheus.Gauge
+	AcquireTotal       prometheus.Counter
+	AcquireFailedTotal prometheus.Counter
+	ReleaseTotal       prometheus.Counter
+}
+
+func newMultiTenantConcurrencyControllerMetrics(reg prometheus.Registerer) *MultiTenantConcurrencyControllerMetrics {
+	return &MultiTenantConcurrencyControllerMetrics{
+		CurrentConcurrency: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_current",
+			Help: "Current number of active global concurrency slots",
+		}),
+		AcquireTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_acquire_total",
```

Comment:
[nit] A comment about metrics naming. "Acquire" and "Release" match the internal implementation but may not be very clear to the final user. I think we can simply call them "started" and "complete". Also "acquire failed" looks like something went bad, but in practice means we were just at capacity; we could rename it too to something less fearful. If you accept this comment, then remember to update help description too.

Reply:
I have used a combination of your suggestion and something else I came up with in https://github.com/grafana/mimir/pull/8146/commits/004f3a11b2208bc5318a49460b1e10fc89d7e3bd - take a look and let me know what you think.